File: Data Structures & Algorithms/invert-a-binary-tree/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# class Solution:
#     def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
#         if not root:
#             return None
        
#         stack = [root]
#         while stack:
#             cur = stack.pop()
#             cur.left, cur.right = cur.right, cur.left
#             if cur.left:
#                 stack.append(cur.left)
            
#             if cur.right:
#                 stack.append(cur.right)
            
#         return root

# class Solution:
#     def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
#         if not root:
#             return None
        
#         que = collections.deque([root])
#         while que:
#             size = len(que)
#             for _ in range(size):
#                 cur = que.popleft()

#                 cur.left, cur.right = cur.right, cur.left
#                 if cur.left:
#                     que.append(cur.left)
                
#                 if cur.right:
#                     que.append(cur.right)
#         return root

# class Solution:
#     def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
#         if not root:
#             return None
        
#         left = self.invertTree(root.left)
#         right = self.invertTree(root.right)
#         root.left = right
#         root.right = left
    
#         return root

class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        self.traverse(root)
        return root
    
    def traverse(self, root):
        if not root:
            return
        
        root.left, root.right = root.right, root.left
        self.traverse(root.left)
        self.traverse(root.right)
    

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-5.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

// class Solution {
//     public int kthSmallest(TreeNode root, int k) {
//         Deque<TreeNode> stack = new ArrayDeque<>();
//         while (root != null || !stack.isEmpty()) {
//             while (root != null) {
//                 stack.push(root);
//                 root = root.left;
//             }
//             TreeNode cur = stack.pop();
//             k--;
//             if (k == 0) {
//                 return cur.val;

//             }
//             root = cur.right;
//         }
//         return -1;
//     }
// }

class Solution {
    private int res = 0;
    private int count = 0;
    public int kthSmallest(TreeNode root, int k) {
        traverse(root, k);
        return res;
    }

    private void traverse(TreeNode root, int k) {
        if (root == null) return;
        traverse(root.left, k);
        count++;
        if (count == k) {
            res = root.val;
            return;
        }
        traverse(root.right, k);
    }
}
